relayer/chains/wasm/provider.go

```diff
@@ -0,0 +1,746 @@
+							time.Sleep(time.Second * 3)
+						} else {
+							blockInfoChan <- &relayTypes.BlockInfo{
+								Height:   height.End,
+								Messages: messages,
+							}
+							break
+						}
+					}
+				}
+			}
+		}
+	}(blockInfoStream, heightStreamChan)
+	return blockInfoStream
+}
+
+func (p *Provider) fetchBlockMessages(ctx context.Context, heightInfo *heightStream) ([]*relayTypes.Message, error) {
+	perPage := 20
+	searchParam := types.TxSearchParam{
+		StartHeight: heightInfo.Start,
+		EndHeight:   heightInfo.End,
+		PerPage:     &perPage,
+	}
+
+	var (
+		wg           sync.WaitGroup
+		messages     coretypes.ResultTxSearch
+		messagesChan = make(chan *coretypes.ResultTxSearch)
+		errorChan    = make(chan error)
+	)
+
+	for _, event := range p.eventList {
+		wg.Add(1)
+		go func(wg *sync.WaitGroup, search types.TxSearchParam, messagesChan chan *coretypes.ResultTxSearch, errorChan chan error) {
+			defer wg.Done()
+			search.Events = append(search.Events, event)
+			res, err := p.client.TxSearch(ctx, search)
+			if err != nil {
+				errorChan <- err
+				return
+			}
+			if res.TotalCount > perPage {
+				for i := 2; i <= int(res.TotalCount/perPage)+1; i++ {
+					search.Page = &i
+					resNext, err := p.client.TxSearch(ctx, searchParam)
+					if err != nil {
+						errorChan <- err
+						continue
+					}
+					res.Txs = append(res.Txs, resNext.Txs...)
+				}
+			}
+			messagesChan <- res
+		}(&wg, searchParam, messagesChan, errorChan)
+		select {
+		case msgs := <-messagesChan:
+			messages.Txs = append(messages.Txs, msgs.Txs...)
+			messages.TotalCount += msgs.TotalCount
+		case err := <-errorChan:
+			return nil, err
+		}
+	}
+	wg.Wait()
+	return p.getMessagesFromTxList(messages.Txs)
+}
+
+func (p *Provider) getMessagesFromTxList(resultTxList []*coreTypes.ResultTx) ([]*relayTypes.Message, error) {
+	var messages []*relayTypes.Message
+	for _, resultTx := range resultTxList {
+		var eventsList []*EventsList
+		if err := json.Unmarshal([]byte(resultTx.TxResult.Log), &eventsList); err != nil {
+			return nil, err
+		}
+
+		for _, event := range eventsList {
+			msgs, err := p.ParseMessageFromEvents(event.Events)
+			if err != nil {
+				return nil, err
+			}
+			for _, msg := range msgs {
+				msg.MessageHeight = uint64(resultTx.Height)
+				p.logger.Info("Detected eventlog",
+					zap.Uint64("height", msg.MessageHeight),
+					zap.String("target_network", msg.Dst),
+					zap.Uint64("sn", msg.Sn),
+					zap.String("event_type", msg.EventType),
+				)
+				messages = append(messages, msg)
+			}
+		}
+	}
+	return messages, nil
+}
+
+func (p *Provider) getRawContractMessage(message *relayTypes.Message) (wasmTypes.RawContractMessage, error) {
+	switch message.EventType {
+	case events.EmitMessage:
+		rcvMsg := types.NewExecRecvMsg(message)
+		return json.Marshal(rcvMsg)
+	case events.CallMessage:
+		execMsg := types.NewExecExecMsg(message)
+		return json.Marshal(execMsg)
+	case events.RevertMessage:
```

Comment:
Are we monitoring this event: `RevertMessage`. I don't see its implementation in the event listener.